cmake/CeleritasLibrary.cmake

```diff
@@ -173,110 +175,138 @@ function(celeritas_add_library target)
     CELERITAS_CUDA_OBJECT_LIBRARY ${target}_objects
   )
+  set_target_properties(${target} PROPERTIES
+    ${_all_props}
+    CELERITAS_CUDA_LIBRARY_TYPE Shared
+    CUDA_RESOLVE_DEVICE_SYMBOLS OFF # We really don't want nvlink called.
+  )
 
-  if(NOT _static_build)
+  ## STATIC ##
+
+  if(_staticsuf)
+    add_library(${target}${_staticsuf} STATIC
+      $<TARGET_OBJECTS:${target}_objects>
+    )
     set_target_properties(${target}${_staticsuf} PROPERTIES
-      LINKER_LANGUAGE CUDA
-      CUDA_SEPARABLE_COMPILATION ON
-      CUDA_RUNTIME_LIBRARY ${_cudaruntime_requested_type}
-      # CUDA_RESOLVE_DEVICE_SYMBOLS OFF # Default for static library
+      ${_all_props}
       CELERITAS_CUDA_LIBRARY_TYPE Static
-      CELERITAS_CUDA_FINAL_LIBRARY ${target}_final
-      CELERITAS_CUDA_MIDDLE_LIBRARY ${target}${_midsuf}
-      CELERITAS_CUDA_STATIC_LIBRARY ${target}${_staticsuf}
-      CELERITAS_CUDA_OBJECT_LIBRARY ${target}_objects
     )
   endif()
 
+  ## FINAL (dlink) ##
+
+  # We need to use a dummy file as a library (per cmake) needs to contains
+  # at least one source file.  The real content of the library will be
+  # the cmake_device_link.o resulting from the execution of `nvcc -dlink`
+  # Also non-cuda related test, for example `gtest_detail_Macros`,
+  # will need to be linked again libceleritas_final while a library
+  # that the depends on and that uses Celeritas::Core (for example
+  # libCeleritasTest.so) will need to be linked against `libceleritas`.
+  # If both the middle and `_final` contains the `.o` files we would
+  # then have duplicated symbols (Here the symptoms will a crash
+  # during the cuda library initialization rather than a link error).
+  celeritas_generate_empty_cu_file(_emptyfilename ${target})
+  add_library(${target}_final ${_lib_requested_type} ${_emptyfilename})
   set_target_properties(${target}_final PROPERTIES
-    LINKER_LANGUAGE CUDA
-    CUDA_RESOLVE_DEVICE_SYMBOLS ON
-    CUDA_SEPARABLE_COMPILATION ON
-    CUDA_RUNTIME_LIBRARY ${_cudaruntime_requested_type}
-    # CUDA_RESOLVE_DEVICE_SYMBOLS ON # Default for shared library
+    ${_all_props}
     CELERITAS_CUDA_LIBRARY_TYPE Final
-    CELERITAS_CUDA_FINAL_LIBRARY ${target}_final
-    CELERITAS_CUDA_STATIC_LIBRARY ${target}${_staticsuf}
-    CELERITAS_CUDA_MIDDLE_LIBRARY ${target}${_midsuf}
-    CELERITAS_CUDA_OBJECT_LIBRARY ${target}_objects
-  )
-
-  target_link_libraries(${target}_final
-    PUBLIC ${target}${_midsuf}
+    CUDA_RESOLVE_DEVICE_SYMBOLS ON
   )
-
+  target_link_libraries(${target}_final PUBLIC ${target})
   target_link_options(${target}_final
-    PRIVATE
-    $<DEVICE_LINK:$<TARGET_FILE:${target}${_staticsuf}>>
+    PRIVATE $<DEVICE_LINK:$<TARGET_FILE:${target}${_staticsuf}>>
   )
-
-  add_dependencies(${target}_final ${target}${_midsuf})
   add_dependencies(${target}_final ${target}${_staticsuf})
 
-  if(_midsuf)
-    add_library(${target} ALIAS ${target}${_midsuf})
+  ## ALIAS/INSTALL ##
+
+  if(PROJECT_NAME STREQUAL "Celeritas")
```

Comment:
I don't quite get what you mean: I agree we want to be in a place where all the cuda linking is done under the hood by some other library, and I want a `celeritas_add_library` call to put the library in the same place as the rest of the code for our project and give it the `Celeritas::` namespace. I have to put the `PROJECT_NAME` exception here because this macro might be used by downstream code.  What we *can* do is rename these functions `cudardc_add_library` (or whatever) and then add a `celeritas_library` wrapper around that which does our project-specific stuff. But I thought that since Ben was working on the genericization we could do what we wanted for now.

Reply:
> But I thought that since Ben was working on the genericization 

I guess it is Ben and I ... and it should still be informed by the bug that we may found in actual usage (i.e. in Celeritas) (one of the commit message says "fix setting of properties of static CUDA library" and it is important that we know if it is a bug fix or a cosmetic fix (or a code cleanup fix, like the factorization of some of the redundant code you did).

So I would indeed go with:

* renaming the existing celeritas_add_library to celeritas_rdc_add_library
* add celeritas_add_library with calls celeritas_rdc_add_library and do the 'PROJECT_NAME exception code' afterward.

Bonus point: once celeritas_rdc_add_library is in VecGeom or some other library or CMake itself, the update will be trivial

Reply:
> fix setting of properties of static CUDA library

1191fefa is indeed a bugfix because the original code had `set(__static_build TRUE)` and `if(NOT _static_build)` (different number of underscores).

OK, I'll separate this out into two separate functions.